kukkuu/tests/test_authentication.py

```diff
@@ -75,11 +98,115 @@ def expired_token_authenticate(*args):
+    guardian = GuardianFactory(email="gustavo.guardian@example.com")
+    response = graphql_request(
+        live_server,
+        headers={
+            "authorization": get_browser_test_bearer_token_for_user(guardian.user)
+        },
+    )
+    assert (
+        response.json()["data"]["myProfile"]["email"] == "gustavo.guardian@example.com"
+    )
+
+
+@patch("kukkuu.oidc.get_or_create_user")
+@patch(
+    "kukkuu.oidc.BrowserTestAwareJWTAuthentication._validate_symmetrically_signed_jwt"
+)
+@pytest.mark.django_db()
+def test_authenticate_test_user(mock_validate_jwt, mock_get_or_create_user):
+    """The test JWT validation should be called when a test token is used."""
+    jwt_claims = {
+        "iss": "test_issuer",
+        "sub": "test_user_id",
+    }
+    jwt = Mock(spec=JWT, claims=jwt_claims)
+    mock_get_or_create_user.return_value = "test_user"
+    auth = BrowserTestAwareJWTAuthentication()
+    result = auth.authenticate_test_user(jwt)
+    assert isinstance(result, UserAuthorization)
+    assert result.user == "test_user"
+    mock_validate_jwt.assert_called_once_with(jwt)
+
+
+def test_get_auth_header_jwt_valid(request_factory):
+    request = request_factory()
+    auth = BrowserTestAwareJWTAuthentication()
+    jwt = auth._get_auth_header_jwt(request)
+    assert isinstance(jwt, JWT)
+
+
+def test_get_auth_header_jwt_invalid_scheme(request_factory):
+    request = request_factory("Basic invalid")
+    auth = BrowserTestAwareJWTAuthentication()
+    jwt = auth._get_auth_header_jwt(request)
+    assert jwt is None
+
+
+def test_get_auth_header_jwt_invalid_bearer(request_factory):
+    request = request_factory("bearer invalid.jwt.structure")
+    auth = BrowserTestAwareJWTAuthentication()
+    with pytest.raises(JWTError):
+        auth._get_auth_header_jwt(request)
+
+
+def test_get_auth_header_jwt_no_header(request_factory):
+    request = request_factory(None)
+    auth = BrowserTestAwareJWTAuthentication()
+    jwt = auth._get_auth_header_jwt(request)
+    assert jwt is None
+
+
+def test_browser_test_auth_enabled_without_issuer_should_raise(settings):
+    """Issuer is a mandatory config when test auth is enabled."""
+    settings.OIDC_BROWSER_TEST_API_TOKEN_AUTH["ENABLED"] = True
+    settings.OIDC_BROWSER_TEST_API_TOKEN_AUTH["ISSUER"] = None
+    with pytest.raises(ImproperlyConfigured):
+        BrowserTestAwareJWTAuthentication()
+
+
+@patch("helusers.oidc.RequestJWTAuthentication.authenticate")
+def test_browser_test_auth_disabled_should_always_call_helusers_authenticate(
+    mock_helusers_authenticate,
+    settings,
+    request_factory,
+):
+    """When browser test authentication is disabled,
+    the helusers authentication is used for test JWTs.
+    """
+    settings.OIDC_BROWSER_TEST_API_TOKEN_AUTH["ENABLED"] = False
+    request = request_factory()
+    auth = BrowserTestAwareJWTAuthentication()
+    auth.authenticate(request)
+    mock_helusers_authenticate.assert_called_once_with(request)
```

Comment:
Would a similar test using `auth.authenticate(request)` but with a test JWT token be good? The test case `test_authenticate_test_user` calls `auth.authenticate_test_user` instead of `auth.authenticate` and none of the other tests seem to test this case.

Reply:
Hmm.. well `test_browser_test_auth_with_expired_token` uses the `auth.authenticate` with test JWT token... so how would you feel about adding a `test_browser_test_auth_with_unexpired_token` test case or similar?